src/Microsoft.Android.Sdk.ILLink/TypeMappingStep.cs

```diff
@@ -15,10 +19,27 @@ namespace Microsoft.Android.Sdk.ILLink;
 public class TypeMappingStep : BaseStep
 {
 	const string AssemblyName = "Microsoft.Android.Runtime.NativeAOT";
-	const string TypeName = "Microsoft.Android.Runtime.NativeAotTypeManager";
+	const string TypeName = "Microsoft.Android.Runtime.TypeMapping";
+	const string SystemIOHashingAssemblyPathCustomData = "SystemIOHashingAssemblyPath";
 	readonly IDictionary<string, List<TypeDefinition>> TypeMappings = new Dictionary<string, List<TypeDefinition>> (StringComparer.Ordinal);
 	AssemblyDefinition? MicrosoftAndroidRuntimeNativeAot;
 
+	AssemblyLoadContext? _hashingAssemblyLoadContext;
+	delegate ulong HashMethod (ReadOnlySpan<byte> data, long seed = 0);
+	HashMethod? _hashMethod;
+
+	protected override void Process ()
+	{
+		if (!Context.TryGetCustomData (SystemIOHashingAssemblyPathCustomData, out var assemblyPath)) {
+			throw new InvalidOperationException ($"The {nameof (TypeMappingStep)} step requires setting the '{SystemIOHashingAssemblyPathCustomData}' custom data");
+		} else if (!System.IO.File.Exists (assemblyPath)) {
+			throw new InvalidOperationException ($"The '{SystemIOHashingAssemblyPathCustomData}' custom data must point to a valid assembly path ('{assemblyPath}' does not exist)");
+		}
+
+		_hashingAssemblyLoadContext = new AssemblyLoadContext (name: null, isCollectible: true);
+		_hashMethod = GetHashMethod (_hashingAssemblyLoadContext, assemblyPath);
```

Comment:
So, this is because we had trouble getting illink to load an assembly reference of a trimmer step?

Was the problem, just getting it to load? Or is a *different* `System.IO.Hashing.dll` already loaded in the process and being used during illink?

Reply:
> Was the problem, just getting it to load?

Yes, the problem was that it wouldn't load it _at all_. I tried copying the assembly into several locations, where I expected the assembly loader to look for the dependencies, but I couldn't get it to load.

ILLink uses [`AssemblyLoadContext.LoadFromAssemblyPath` internally](https://github.com/dotnet/runtime/blob/main/src/tools/illink/src/linker/Linker/Driver.cs#L910) and it appears that it won't load other referenced assemblies if they aren't included in the trimmer's `deps.json` file?

@sbomer do you have any suggestions how to make sure the `System.IO.Hashing.dll` dependency of our custom liker step is loaded without doing it automatically?

Reply:
I wondered if it would load if you add the assembly to `@(_TrimmerCustomSteps)` with no metadata:

https://github.com/dotnet/android/blob/60eb9de435e2857c47465148a074a144e9e6eaf4/src/Xamarin.Android.Build.Tasks/Microsoft.Android.Sdk/targets/Microsoft.Android.Sdk.ILLink.targets#L47

Reply:
Which `System.IO.Hashing.dll` is being loaded here? Is it the netstandard2.0 one?
When building in VS, does the Linker run in a dotnet environment or a full framework msbuild?

Reply:
never mind, looks like our Tasks already reference `System.IO.Hashing` so it will be the netstandard2.0 one